### ISSUE TITLE
NoOptionError while initial configuration.

### ISSUE THREAD

**Issue by n0tty** [sopel-irc/sopel#999]
Hey,
Post installation, when I tried 1st run, I got the following error -
```
Welcome to Sopel!
I can't seem to find the configuration file, so let's generate it!

Please answer the following questions to create your configuration file:

Encountered an error while writing the config file. This shouldn't happen. Check permissions.
Traceback (most recent call last):
  File "/usr/local/bin/sopel", line 9, in <module>
    load_entry_point('sopel==6.1.1', 'console_scripts', 'sopel')()
  File "/usr/local/lib/python2.7/dist-packages/sopel-6.1.1-py2.7.egg/sopel/run_script.py", line 132, in main
    _create_config(configpath)
  File "/usr/local/lib/python2.7/dist-packages/sopel-6.1.1-py2.7.egg/sopel/config/__init__.py", line 251, in _create_config
    config = Config(configpath, validate=False)
  File "/usr/local/lib/python2.7/dist-packages/sopel-6.1.1-py2.7.egg/sopel/config/__init__.py", line 63, in __init__
    validate=validate)
  File "/usr/local/lib/python2.7/dist-packages/sopel-6.1.1-py2.7.egg/sopel/config/__init__.py", line 119, in define_section
    setattr(self, name, cls_(self, name, validate=validate))
  File "/usr/local/lib/python2.7/dist-packages/sopel-6.1.1-py2.7.egg/sopel/config/types.py", line 59, in __init__
    getattr(self, value)
  File "/usr/local/lib/python2.7/dist-packages/sopel-6.1.1-py2.7.egg/sopel/config/types.py", line 147, in __get__
    value = instance._parser.get(instance._section_name, self.name)
  File "/usr/lib/python2.7/ConfigParser.py", line 340, in get
    raise NoOptionError(option, section)
ConfigParser.NoOptionError: No option u'admins' in section: u'core'
```
I also edited the run_script.py to allow root user to use the same and retried it from the root user, but still got the same error.

**Comment by dasu** [comment]
Does ~/.sopel/ exist? 
If so,  what are the permissions of that folder and default.cfg ?

**Comment by n0tty** [comment]
~/.sopel exists on a non-root user and root user with perms 755

**Comment by maxpowa** [comment]
@n0tty Which version of python are you running?

**Comment by maxpowa** [comment]
So somehow, I have no clue how, an except block is not catching the exception. We have code that intercepts this error ([146-155 in types.py](https://github.com/sopel-irc/sopel/blob/664d9f444c3086a9a7bf150efbe24b281dc816ea/sopel/config/types.py#L146-L155)) but for whatever reason it's not working. Perhaps do you have another (non-corelib) `configparser` package installed?

**Comment by n0tty** [comment]
~$ python --version
Python 2.7.9

@maxpowa Nothing different from the standard python configparser. However, I am gonna fiddle around with this later to check out what can be done. Hadn't had much time to play around, will play around and let you know the exactly problem by next week if I can.

**Comment by embolalia** [comment]
What command are you running, exactly? How did you install Sopel? I can't reproduce this on python 2 or 3. I can get different exceptions by creating an incomplete config file manually and starting Sopel with that, but it doesn't attempt to run the wizard like you're seeing.

**Comment by n0tty** [comment]
@embolalia 
$pip install sopel
Also built it from source.

**Comment by n0tty** [comment]
Hi. So I just was debugging the code. I found out that it was something to do with the permissions. Currently I commented out the line 93 and 98 lines from run_script.py so that it would run as root. After this once I build the installation, the execution fails at line 132 _create_config(configpath) stating Permission Error. So I chown -R root:root * the sopel directory. However, it still gives the same error.

**Comment by elad661** [comment]
You mustn't run sopel as root. You need to create a user for it and run it as that user. Since that user will have read/write permissions to its own home folder, sopel will be able to create the .sopel directory properly.

This means your bug report is invalid, because you modified sopel to remove a check that has to be there for security and sanity reasons.

@embolalia can I close this issue? Or is there something of value here that I'm missing?

**Comment by n0tty** [comment]
@elad661 I did run this as a non root user as well.
I am using Kali Linux 2.0. Non-root user. I switched to root and modified the script only because it wasn't working as non-root. Since it was permission problem, I did the modification and run as root.

Also, you can replicate this bug if you download a Kali Linux VM and try it on that. I tested the same out on a different machine. But same issue.

**Comment by elad661** [comment]
Look, Kali is designed to be run of a live usb to do pen testing, it's not designed for general use, so I wouldn't recommend using it for anything other than pen testing. With that said,

There's no real reason there will be a permission error there. If you got a permission error when running Sopel in a user you just created, it means that your user either has no home directory, or no permissions within it's own home directory. It might mean your distro is incompetent in that the useradd command doesn't create the user's home directory, or that you added a user the wrong way.

Your initial comment claimed that `~/.sopel` exists with the correct permissions on a non-root  user, but then you mentioned you ran it as root. This makes no sense to me. You also said you changed the directory ownership of ~/.sopel to `root`, if you do that and then run sopel as a normal user, of course you'd get permission issues.

I've tried to reproduce your issue by manually removing stuff from my config file, and it still works. So, I don't see how you'd get Sopel to fail like this.

With that said, there is an option to run Sopel with a user that has no home directory, and you can see an example for this in the systemd service and default sopel config file in `contrib/`

You might have a better luck running a general-purpose or a desktop distribution, such as Fedora, Debian, Ubuntu, RHEL, CentOS, Mint, Arch or something like that...

But since I feel bad about just dismissing your bug as your OS doing something weird, please follow the following steps:

1. DO NOT modify the sopel core to remove sanity checks. Please reinstall Sopel with all the original sanity checks untouched.
2. Create a new user specifically for this test (different than the one you created before), and make sure it has a home directory and permissions to its own home directory. On a sane distro, this is the default. 
3. Switch to this user using `su -` (the minus is important to get a full login environment)
4. Run sopel within that user.

Then, if it fails, please provide the output of the following commands (without exiting that user you created):

`ls -RlZa ~` - so we can see the permission and selinux context of all files in that user's home dir. (If you create a new user specifically for this test like I asked, there won't be any private information there, which is why I'm comfortable with asking for it with the R parameter)

`getenforce` - so we can see if SELinux is installed and if it's in enforcing mode

`stat ~` - so we can see the permissions and SELinux context on the home directory itself

This will give us something to go on.

**Comment by n0tty** [comment]
I **also** edited the run_script.py to allow root user to use the same and retried it from the root user, but still got the same error.
By **also** I mean, I first ran this as a non-root user, as recommended practice. As it did not work, I bypassed the checks and **also** ran it as a root user.

As this is my office laptop, I can not clean install the OS(s) that you mentioned, and obviously that is the reason I do most of my day to day activities from Kali linux which work just fine. That being said, I donno why Kali Linux would prevent it from getting installed and wanted to know what exactly caused this. And as you guys have a lot of exp in building this I thought I would ask it out here. Thanks.

**Comment by elad661** [comment]
If you don't provide the debugging information I asked for, there's no way for us to help you, as we can't reproduce the problem ourselves.

**Comment by leifmadsen** [comment]
I get this same issue when installing sopel on Fedora 23. To reproduce I did:

```
pip install --user sopel
sopel -w
```

Python version: 2.7.11
Sopel version: 6.3.1

I have SElinux set to enforcing, but I've tried with Permissive and get the same issue.

```
stat ~
  File: ‘/home/lmadsen’
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: fd03h/64771d	Inode: 11010049    Links: 83
Access: (0711/drwx--x--x)  Uid: ( 1000/ lmadsen)   Gid: ( 1000/ lmadsen)
Context: unconfined_u:object_r:user_home_dir_t:s0
Access: 2016-06-10 15:44:10.616829095 -0400
Modify: 2016-06-10 15:43:45.430270908 -0400
Change: 2016-06-10 15:43:45.430270908 -0400
 Birth: -
```

Returning output of `ls -RlZa ~` is far too verbose for this issue, but here it is for the `.sopel` dir:

```
ls -RlZa ~/.sopel
/home/lmadsen/.sopel:
total 8
drwxrwxr-x.  2 lmadsen lmadsen unconfined_u:object_r:user_home_t:s0     4096 Jun 10 15:43 .
drwx--x--x. 83 lmadsen lmadsen unconfined_u:object_r:user_home_dir_t:s0 4096 Jun 10 15:43 ..
```

Here is a run with things in Permissive mode:

```
$ sudo setenforce 0
$ getenforce
Permissive
$ sopel -w
Please answer the following questions to create your configuration file:

Encountered an error while writing the config file. This shouldn't happen. Check permissions.
Traceback (most recent call last):
  File "/home/lmadsen/.local/bin/sopel", line 9, in <module>
    load_entry_point('sopel==6.3.1', 'console_scripts', 'sopel')()
  File "/home/lmadsen/.local/lib/python2.7/site-packages/sopel/run_script.py", line 108, in main
    _wizard('all', opts.config)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/sopel/config/__init__.py", line 219, in _wizard
    _create_config(configpath)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/sopel/config/__init__.py", line 252, in _create_config
    config = Config(configpath, validate=False)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/sopel/config/__init__.py", line 64, in __init__
    validate=validate)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/sopel/config/__init__.py", line 120, in define_section
    setattr(self, name, cls_(self, name, validate=validate))
  File "/home/lmadsen/.local/lib/python2.7/site-packages/sopel/config/types.py", line 59, in __init__
    getattr(self, value)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/sopel/config/types.py", line 147, in __get__
    value = instance._parser.get(instance._section_name, self.name)
  File "/usr/lib64/python2.7/ConfigParser.py", line 340, in get
    raise NoOptionError(option, section)
ConfigParser.NoOptionError: No option u'admin_accounts' in section: u'core'
```

**Comment by eanxgeek** [comment]
I am going to chime in here as well, I have a Fedora 23 instance with Python 2.7.10 installed.  This system has not been updated recently and is working perfectly with sopel 6.3.1.

I have another system running Fedora 24, it was running perfectly until I ran 'dnf update' yesterday.  Today I get this exact error.

I have already tried to rollback my python-libs package, that did not help.  I have also turned off selinux and enabled full write permission with no change.

I'm now at the point where I am going to clone my working system, then update it and see if it breaks.
